lib/make-disk-image.nix

```diff
@@ -24,7 +24,14 @@ let
     util-linux
   ];
   preVM = ''
-    ${lib.concatMapStringsSep "\n" (disk: "truncate -s ${disk.imageSize} ${disk.name}.raw") (lib.attrValues nixosConfig.config.disko.devices.disk)}
+    export PATH=$PATH:${lib.makeBinPath (with pkgs; [ nix jq ])}
+    export NIX_STATE_DIR=$TMPDIR/state
+    export NIX_CONFIG="experimental-features = nix-command"
+    nix-store --load-db < ${pkgs.closureInfo {
+      rootPaths = [ systemToInstall.config.system.build.toplevel ];
+    }}/registration
+
+    ${lib.concatMapStringsSep "\n" (disk: "truncate -s $(($(set -x; nix path-info --json ${systemToInstall.config.system.build.toplevel} | jq .[].closureSize) + ${toString disk.imageSizeExtraBytes})) ${disk.name}.raw") (lib.attrValues nixosConfig.config.disko.devices.disk)}
```

Comment:
newlines could help to make that more readable. and probably using '' instead of "

Reply:
I would also assign variables so that in the event of a failure it's easier to spot which part failed.